toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1073 @@
+    // to map directory names to lists of nsMainThreadPtr<callbacks>.
+    nsClassHashtable<nsStringHashKey, ChangeCallbackArray> mChangeCallbacksTable;
+    nsClassHashtable<nsStringHashKey, ErrorCallbackArray> mErrorCallbacksTable;
+
+    // Other methods need to know that a shutdown is in progress.
+    bool mShuttingDown;
+
+    // Main inotify file descriptor (initialized in NativeFileWatcher's Init())
+    int mInotifyFileDescriptor;
+
+    // Queue for the events read from inotify file descriptor
+    std::queue<inotify_event*> mInotifyEventQueue;
+
+    nsresult RunInternal();
+
+    nsresult DispatchChangeCallbacks(
+            WatchedResourceDescriptor* aResourceDescriptor,
+            const nsAString& aChangedResource);
+
+    nsresult ReportChange(
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsAString& aChangedResource);
+
+    nsresult DispatchErrorCallbacks(
+            WatchedResourceDescriptor* aResourceDescriptor,
+            nsresult anError,
+            long anOSError);
+
+    nsresult ReportError(
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+            nsresult anError,
+            long anOSError);
+
+    nsresult ReportSuccess(const nsMainThreadPtrHandle<
+                           nsINativeFileWatcherSuccessCallback>& aOnSuccess,
+                           const nsAString& aResourcePath);
+
+    void AppendCallbacksToHashtables(
+            const nsAString& aPath,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError);
+
+    void RemoveCallbacksFromHashtables(
+            const nsAString& aPath,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError);
+
+    nsresult MakeResourcePath(WatchedResourceDescriptor* changedDescriptor,
+                              const nsAString& resourceName,
+                              nsAString& nativeResourcePath);
+};
+
+/**
+ * The watching thread logic.
+ *
+ * @return NS_OK if the watcher loop must be rescheduled, a failure code
+ *         if it must not.
+ */
+nsresult
+NativeFileWatcherIOTask::RunInternal()
+{
+    int mInotifyReadTimeout;
+    char inotifyEventTempBuffer[(sizeof(struct inotify_event) + NAME_MAX + 1)] __attribute__ ((aligned(8)));
+
+    // Sleep to keep thread in suspended state until event occurs, and to keep thread from using CPU.
+    // Will return early as soon as an event occurs (our signal handler is triggered).
+    sleep(0.01);
+
+    // If we've slept for 0.01 seconds (or we returned early due to a file system event), we should check
+    // for events waiting to be read from the file descriptor.
+    if(mEventWaiting){
+        --mEventWaiting; // Decrement an event, since we're acting on one of them.
```

Comment:
I suspect that you're not decrementing *an event*, but rather *the number of pending events*.

Also, `sig_atomic_t` protects you from interactions between the signal handler and the main thread but **not** from interactions between the signal handler and your I/O thread, so `sig_atomic_t` won't work here.

Real atomics would probably work (I'll admit I haven't checked how they interact with signals), see [mfbt/Atomics.h](https://searchfox.org/mozilla-central/source/mfbt/Atomics.h).

Or, you know, get rid of signals altogether :)